test/functional/session_controller_test.rb

```diff
@@ -0,0 +1,118 @@
+      stub_account_api_validates_auth_response
+      get :callback, params: { state: "abc123" }
+
+      assert_response :redirect
+      assert_equal @response.redirect_url, @root_path
+    end
+
+    should "redirect to the root path if the :state param is not present" do
+      stub_account_api_validates_auth_response
+      get :callback, params: { code: "abc123" }
+
+      assert_response :redirect
+      assert_equal @response.redirect_url, @root_path
+    end
+
+    should "set the 'GOVUK-Account-Session' header" do
+      stub_account_api_validates_auth_response(govuk_account_session: @govuk_account_session)
+      get :callback, params: { code: "abc123", state: "abc123" }
+
+      assert_equal @response.headers["GOVUK-Account-Session"], @govuk_account_session
+    end
+
+    should "redirect to the redirect path if a redirect path is in the account-api response" do
+      stub_account_api_validates_auth_response(redirect_path: "/bank-holiday")
+      get :callback, params: { code: "abc123", state: "abc123" }
+
+      assert_response :redirect
+      assert_includes @response.redirect_url, "/bank-holiday"
+    end
+
+    should "redirect to the account manager url if no redirect path is in the account-api response" do
+      stub_account_api_validates_auth_response(redirect_path: nil, ga_client_id: nil)
+      get :callback, params: { code: "abc123", state: "abc123" }
+
+      assert_response :redirect
+      assert_equal @response.redirect_url, Plek.find("account-manager")
+    end
+
+    should "redirect with the specified :ga_client_id of the api response if present" do
+      ga_client_id = "ga_client_id"
+      stub_account_api_validates_auth_response(ga_client_id: ga_client_id)
+      get :callback, params: { code: "abc123", state: "abc123" }
+
+      assert_response :redirect
+      assert_includes @response.redirect_url, ga_client_id
+    end
+
+    should "redirect with the specified :_ga param if :ga_client_id of the api response is not present" do
+      underscore_ga = "underscore_ga"
+      stub_account_api_validates_auth_response(ga_client_id: nil)
+      get :callback, params: { code: "abc123", state: "abc123", _ga: underscore_ga }
+
+      assert_response :redirect
+      assert_includes @response.redirect_url, underscore_ga
+    end
+
+    should "respond with :bad_request if :code or :state are invalid" do
+      stub_account_api_rejects_auth_response
+      get :callback, params: { code: "abc123", state: "abc123" }
+
+      assert_response :bad_request
+    end
+  end
+
+  context "GET /sign-out" do
+    context "when the user is logged in" do
+      setup do
+        mock_logged_in_session
+        @account_end_session_header_name = AccountConcern::ACCOUNT_END_SESSION_HEADER_NAME
+      end
+
+      should "set ACCOUNT_END_SESSION_HEADER_NAME response header to 1" do
+        get :delete
+        assert_includes @response.headers[@account_end_session_header_name], "1"
```

Comment:
Use `AccountConcern::ACCOUNT_END_SESSION_HEADER_NAME` directly here, instead of assigning it to an instance variable first.

Reply:
Also the actual value of the header is unimportant, only that there is one.  So this test could just check that the header is present.

Reply:
ace